packages/react-storage/src/components/StorageBrowser/views/LocationDetailView/useLocationDetailView.ts

```diff
@@ -0,0 +1,156 @@
+  onRefresh: () => void;
+  onPaginateNext: () => void;
+  onPaginatePrevious: () => void;
+  onAddFiles: (files: File[]) => void;
+}
+
+export type LocationDetailViewActionType =
+  | { type: 'REFRESH_DATA' } // refresh data only
+  | { type: 'RESET' } // reset view to initial state
+  | { type: 'PAGINATE_NEXT' }
+  | { type: 'PAGINATE_PREVIOUS' }
+  | { type: 'PAGINATE'; page: number }
+  | { type: 'ACCESS_ITEM'; key: string }
+  | { type: 'NAVIGATE'; location: LocationData }
+  | { type: 'ADD_FILES'; files: File[] }
+  | { type: 'SEARCH'; query: string; includeSubfolders?: boolean };
+
+interface InitialValues {
+  pageSize?: number;
+  delimiter?: string;
+}
+
+export interface UseLocationDetailViewOptions {
+  initialValues?: InitialValues;
+  onDispatch?: React.Dispatch<LocationDetailViewActionType>;
+  onActionSelect?: (type: string) => void;
+  onExit?: () => void;
+  onNavigate?: (destination: LocationData) => void;
+}
+
+const DEFAULT_PAGE_SIZE = 100;
+export const DEFAULT_LIST_OPTIONS = {
+  delimiter: '/',
+  pageSize: DEFAULT_PAGE_SIZE,
+};
+
+export function useLocationDetailView(
+  options?: UseLocationDetailViewOptions
+): UseLocationDetailView {
+  const initialValues = options?.initialValues;
+  const onActionSelect = options?.onActionSelect;
+  const onNavigate = options?.onNavigate;
+  const listOptions = React.useMemo(() => {
+    return {
+      ...DEFAULT_LIST_OPTIONS,
+      ...initialValues,
+    };
+  }, [initialValues]);
+
+  const [{ history }, dispatchStoreAction] = useStore();
+  const { current } = history;
+  const { prefix } = current ?? {};
+  const hasInvalidPrefix = isUndefined(prefix);
+
+  const [{ data, isLoading, hasError, message }, handleList] = useAction(
+    'LIST_LOCATION_ITEMS'
+  );
+
+  const { result, nextToken } = data;
+  const resultCount = result.length;
+  const hasNextToken = !!nextToken;
+  const onPaginateNext = () => {
+    if (hasInvalidPrefix || !nextToken) return;
+    dispatchStoreAction({ type: 'RESET_LOCATION_ITEMS' });
+    handleList({ prefix, options: { ...listOptions, nextToken } });
+  };
+
+  const onPaginatePrevious = () => {
+    dispatchStoreAction({ type: 'RESET_LOCATION_ITEMS' });
+  };
+
+  const {
+    currentPage,
+    handlePaginateNext,
+    handlePaginatePrevious,
+    handleReset,
+    range,
+  } = usePaginate({
+    onPaginateNext,
+    onPaginatePrevious,
+    pageSize: listOptions.pageSize,
+  });
+
+  const onRefresh = () => {
+    if (hasInvalidPrefix) return;
+    handleReset();
+    handleList({ prefix, options: { ...listOptions, refresh: true } });
+    dispatchStoreAction({ type: 'RESET_LOCATION_ITEMS' });
+  };
+
+  React.useEffect(() => {
+    if (hasInvalidPrefix) return;
+    handleList({ prefix, options: { ...listOptions, refresh: true } });
+    handleReset();
+  }, [
+    dispatchStoreAction,
+    handleList,
+    handleReset,
+    listOptions,
+    hasInvalidPrefix,
+    prefix,
+  ]);
+
+  const processedItems = React.useMemo(() => {
```

Comment:
Nit: Might as well just call this `pageItems` and use shorthand below since "processed" doesn't really add much context anyway.